File: examples/quickstart/python/ai_utils.py
import os
import openai;
import logging;
from typing import List
from enum import Enum
from models import Todo

logger = logging.getLogger(__name__)
# Wrapper functions around embedding generation and chat models

class EmbeddingTasks(Enum):
    SEARCH_DOCUMENT = "search_document:"
    SEARCH_QUERY = "search_query:"

# Nomic models are fine tuned for specific tasks. They benefit from a prefix that describes the task:
def adjust_input(text: str, task: EmbeddingTasks) -> str:
    if ("nomic" in os.getenv("EMBEDDING_MODEL")):
        return task.value + text
    else:
        return text

def get_embedding(text: str, task: EmbeddingTasks) -> List[float]:
    client = openai.OpenAI(
        base_url = os.getenv("AI_BASE_URL"),
        api_key=os.getenv("AI_API_KEY"),
    )
        
    response = client.embeddings.create(
        model=os.getenv("EMBEDDING_MODEL"),
        input=adjust_input(text, task),
        # dimensions=128, # optional, nomic models are 768 by default, but scale down
    )
    
    return response.data[0].embedding

def get_similar_tasks(session: any, text: str):
    query_embedding = get_embedding(text, EmbeddingTasks.SEARCH_QUERY)
    similar_tasks_raw = (
        session.query(Todo)
        .filter(Todo.embedding.cosine_distance(query_embedding) < 1)
        .order_by(Todo.embedding.cosine_distance(query_embedding)).limit(3))
    return [{"title": task.title, "estimate": task.estimate} for task in similar_tasks_raw]

def ai_estimate(text: str, similar_tasks):
    client = openai.OpenAI(
        base_url = os.getenv("AI_BASE_URL"),
        api_key = os.getenv("AI_API_KEY"),
    )
    
    response = client.chat.completions.create(
        model = os.getenv("AI_MODEL"),
        messages = [
            {
                "role": "user",
                "content" : 
                    f'you are an amazing project manager. I need to {text}. How long do you think this will take?' 
                    f'I have a few similar tasks with their estimates, please use them as reference: {similar_tasks}.'
                    f'respond with just the estimate, no yapping.',
             },
        ],   
    )
    
    # if we got a valid response, return it
    if (response.choices[0].finish_reason == "stop"):
        return response.choices[0].message.content
    else:
        # otherwise, we simply don't have an estimate
        return "unknown"
    
    
    
    
    

File: examples/quickstart/node_react/src/be/AiUtils.ts
import { OpenAI } from "openai";
import { Server } from "@niledatabase/server";

export enum EmbeddingTasks {
  SEARCH_DOCUMENT = "search_document:",
  SEARCH_QUERY = "search_query:",
}

export interface todo {
  title: string;
  estimate: string;
}

const DEFAULT_MODEL = "nomic-ai/nomic-embed-text-v1.5";
const EMBEDDING_MODEL = process.env.EMBEDDING_MODEL || DEFAULT_MODEL;

function adjust_input(text: string, task: EmbeddingTasks): string {
  if (EMBEDDING_MODEL?.indexOf("nomic") >= 0) {
    return task + text;
  } else {
    return text;
  }
}

export function embeddingToSQL(embedding: number[]) {
  return JSON.stringify(embedding);
}

export async function embedTask(title: string, task: EmbeddingTasks) {
  console.log(`Embedding task: ${title} ${task}`);
  const ai = new OpenAI({
    apiKey: process.env.AI_API_KEY,
    baseURL: process.env.AI_BASE_URL,
  });

  // generate embeddings
  let resp = await ai.embeddings.create({
    model: EMBEDDING_MODEL,
    input: title,
    dimensions: 768,
  });

  // OpenAI's response is an object with an array of
  // objects that contain the vector embeddings
  // We just return the vector embeddings
  const embedding = resp.data[0].embedding;
  console.log(`Embedding dimensions: ${embedding.length}`);
  return embedding;
}

export async function findSimilarTasks(
  tenantNile: Server,
  title: string
): Promise<todo[]> {
  const embedding = await embedTask(title, EmbeddingTasks.SEARCH_QUERY);

  // get similar tasks, no need to filter by tenant because we are already in the tenant context
  const similarTasks = await tenantNile.db.query(
    `SELECT title, estimate FROM todos WHERE embedding <-> $1 < 1 ORDER BY embedding <-> $1 LIMIT 3;`,
    [embeddingToSQL(embedding)]
  );

  console.log(` found ${similarTasks.rowCount} similar tasks`);

  return similarTasks.rows;
}

export async function aiEstimate(title: string, similarTasks: todo[]) {
  // I use Fireworks as the model vendor for this example, but any OpenAI-compatible vendor will work
  // Just swap the API key, URL and model
  const ai = new OpenAI({
    apiKey: process.env.AI_API_KEY,
    baseURL: process.env.AI_BASE_URL,
  });

  const model =
    process.env.AI_MODEL ||
    "accounts/fireworks/models/llama-v3p1-405b-instruct";

  const message = `you are an amazing project manager. I need to ${title}. How long do you think this will take?${
    similarTasks.length > 0
      ? ` I have a few similar tasks with their estimates in json format, please use them as reference: ${JSON.stringify(similarTasks.map(task => ({
          title: task.title,
          estimate: task.estimate
        })), null, 2)}.`
      : ''
  }
  respond with just the estimate, keep the answer short.`;

  console.log("Message to AI:", message);

  const aiEstimate = await ai.chat.completions.create({
    messages: [
      {
        role: "user",
        content: message,
      },
    ],
    max_tokens: 64, // limit the response to 64 tokens
    model: model,
  });

  console.log("aiEstimate", aiEstimate);

  // if we got a valid response, return it
  if (aiEstimate.choices[0].finish_reason === "stop") {
    return aiEstimate.choices[0].message.content;
  }
  // otherwise, we simply don't have an estimate
  return "unknown";
}
